scripts/ci/jobs/diff-dumps.sh

```diff
@@ -107,48 +107,88 @@ create_offline_dump() {
-    curl --silent --show-error --max-time 60 --retry 3 --create-dirs -o out/RELEASE_VERSION.txt https://raw.githubusercontent.com/stackrox/stackrox/master/scanner/updater/version/RELEASE_VERSION
-    version_file="out/RELEASE_VERSION.txt"
-
-    latest_version=""
-    # Get the latest X.Y version which has been released.
-    while read -r version; do
-        status_code=$(curl --silent --max-time 60 --retry 3 --write-out "%{http_code}" -o /dev/null "https://github.com/stackrox/stackrox/releases/tag/$version.0")
-        if [[ "$status_code" == "404" ]]; then
-            # Release was not cut, so try a previous release.
-            continue
-        elif [[ "$status_code" == "200" ]]; then
-            # Release was cut, so use this one.
-            latest_version="$version"
-            break
-        else
-            # Some other error occurred.
-            >&2 echo "received status code $status_code when fetching StackRox releases"
-            exit 1
-        fi
-    done < <(grep -oE '^[0-9]+\.[0-9]+' "$version_file" | sort --reverse --unique --version-sort)
-    if [[ -z "$latest_version" ]]; then
-      >&2 echo "programmer error: latest_version unset..."
-      exit 1
-    fi
-
-    file_to_check="scanner-v4-defs-${latest_version}.zip"
-
-    curl --silent --show-error --fail --max-time 60 --retry 3 -o "$file_to_check" "https://definitions.stackrox.io/v4/offline-bundles/$file_to_check"
-    unzip -l "$file_to_check"
-    zip scanner-vuln-updates.zip "$file_to_check"
-    echo "$file_to_check added to scanner-vuln-updates.zip"
+# scanner_v4_build_bundles() creates versioned offline bundles for V4 and add V4
+# definitions to the latest offline bundle.  It determines the version to use
+# for each release based on the VULNERABILITY_VERSION value for each release
+# tag.
+#
+scanner_v4_create_and_add_bundles() {
+    local latest_bundle="${1:-scanner-vuln-updates.zip}"
+
+    local curl="curl --silent --show-error --max-time 60 --retry 3"
+    local release_version_url="https://raw.githubusercontent.com/stackrox/stackrox/master/scanner/updater/version/RELEASE_VERSION"
+    local vuln_version_pattern="https://raw.githubusercontent.com/stackrox/stackrox/%s/scanner/VULNERABILITY_VERSION"
+
+    [ -f scanner-defs.zip ] || die "missing required bundle inclusion: scanner-defs.zip"
+    [ -f k8s-istio.zip ]    || die "missing required bundle inclusion: k8s-istio.zip"
+    [ -f "$latest_bundle" ] || die "missing latest bundle: $latest_bundle"
+
+    info "existing latest bundle:"
+    unzip -l "$latest_bundle"
+
+    info "fetching release versions at $release_version_url"
+
+    # Read the supported releases file, filter lines with release versions,
+    # parse 4.4.* and 4.5.* to print only the X.Y, otherwise print the whole
+    # X.Y.Z, finally merge everything together, and pipe one release per line.
+
+    $curl --fail "$release_version_url" \
+        | grep '^[0-9]\.[0-9]\.[0-9]$' \
+        | sed -En 's/^(4\.4|4\.5)\..*/\1/p;t; p' \
```

Comment:
nit: on mac this produces an error, can it be made more cross-platform friendly?

```
sed: 1: "s/^(4\.4|4\.5)\..*/\1/p ...": undefined label '; p'
```

Reply:
> on mac this produces an error, can it be made more cross-platform friendly?

This script depends on `sed` as shipped by the CI image `quay.io/stackrox-io/apollo-ci:scanner-test-0.4.3`. I mean, it's OK to work it out so you can run this on any `sed` implementation, but I don't think we should try to achieve this considering we can run/test this inside `quay.io/stackrox-io/apollo-ci:scanner-test-0.4.3`.

Saying that, in this case, there is a simpler expression that I believe is supported by multiple implementations (although, I cannot fully assure you as I don't know which one you're using, so I am hoping for the best).

92a330e0

Reply:
Makes sense! Didn't realize these were run inside `quay.io/stackrox-io/apollo-ci:scanner-test-0.4.3`. In that case is/was OK.